org.jacoco.core/src/org/jacoco/core/internal/analysis/Instruction.java

```diff
@@ -160,22 +159,27 @@ public Instruction merge(final Instruction other) {
 
 	/**
 	 * Creates a copy of this instruction where all outgoing branches are
-	 * replaced with the given instructions. The coverage status of the new
-	 * instruction is derived from the status of the given instructions.
+	 * replaced. The coverage status of the branches of the new instruction is
+	 * derived from the status of the given branches of the given instructions.
 	 *
-	 * @param newBranches
-	 *            new branches to consider
+	 * @param branches
+	 *            indexes of new branches whose execution status should be
+	 *            computed
+	 * @param fromInstructions
+	 *            instructions whose branch execution statuses should be used
+	 * @param fromBranches
+	 *            indexes of branches of the given instructions whose execution
+	 *            status should be used
 	 * @return new instance with replaced branches
 	 */
-	public Instruction replaceBranches(
-			final Collection<Instruction> newBranches) {
+	public Instruction replaceBranches(final int[] branches,
+			final Instruction[] fromInstructions, final int[] fromBranches) {
 		final Instruction result = new Instruction(this.line);
-		result.branches = newBranches.size();
-		int idx = 0;
-		for (final Instruction b : newBranches) {
-			if (!b.coveredBranches.isEmpty()) {
-				result.coveredBranches.set(idx++);
+		for (int i = 0; i < branches.length; i++) {
+			if (fromInstructions[i].coveredBranches.get(fromBranches[i])) {
+				result.coveredBranches.set(branches[i]);
 			}
+			result.branches = Math.max(result.branches, branches[i] + 1);
```

Comment:
As a side note:

`Math.max` is used here because some unit tests violate contract of `Instruction` that states that after CFG construction [each instruction has at least one branch](https://github.com/jacoco/jacoco/blob/e6f29c902242c24c20f7740e27006a83330b8e88/org.jacoco.core/src/org/jacoco/core/internal/analysis/Instruction.java#L23).